helix-core/src/history.rs

```diff
@@ -22,72 +58,195 @@ impl Default for History {
+        up_txns.chain(down_txns).collect()
+    }
+
+    fn jump_backward(&mut self, delta: usize) -> Vec<Transaction> {
+        self.jump_to(self.current.saturating_sub(delta))
+    }
+
+    fn jump_forward(&mut self, delta: usize) -> Vec<Transaction> {
+        use std::cmp::max;
+        self.jump_to(max(
+            self.revisions.len() - 1,
+            self.current.saturating_add(delta),
+        ))
+    }
 
-            return Some(&transaction);
+    // Helper for a binary search case below.
+    fn revision_closer_to_instant(&self, i: usize, instant: Instant) -> usize {
+        let dur_im1 = instant.duration_since(self.revisions[i - 1].timestamp);
+        let dur_i = self.revisions[i].timestamp.duration_since(instant);
+        use std::cmp::Ordering::*;
+        match dur_im1.cmp(&dur_i) {
+            Less => i - 1,
+            Equal | Greater => i,
+        }
+    }
+
+    fn jump_instant(&mut self, instant: Instant) -> Vec<Transaction> {
+        let search_result = self
+            .revisions
+            .binary_search_by(|rev| rev.timestamp.cmp(&instant));
+        let revision = match search_result {
+            Ok(revision) => revision,
+            Err(insert_point) => match insert_point {
+                0 => 0,
+                n if n == self.revisions.len() => n - 1,
+                i => self.revision_closer_to_instant(i, instant),
+            },
+        };
+        self.jump_to(revision)
+    }
+
+    fn jump_duration_backward(&mut self, duration: Duration) -> Vec<Transaction> {
+        match self.revisions[self.current].timestamp.checked_sub(duration) {
+            Some(instant) => self.jump_instant(instant),
+            None => self.jump_to(0),
+        }
+    }
+
+    fn jump_duration_forward(&mut self, duration: Duration) -> Vec<Transaction> {
+        match self.revisions[self.current].timestamp.checked_add(duration) {
+            Some(instant) => self.jump_instant(instant),
+            None => self.jump_to(self.revisions.len() - 1),
+        }
+    }
+
+    pub fn earlier(&mut self, sotp: StepsOrTimePeriod) -> Vec<Transaction> {
+        use StepsOrTimePeriod::*;
+        match sotp {
+            Steps(n) => self.jump_backward(n),
+            TimePeriod(d) => self.jump_duration_backward(d),
+        }
+    }
+
+    pub fn later(&mut self, sotp: StepsOrTimePeriod) -> Vec<Transaction> {
+        use StepsOrTimePeriod::*;
+        match sotp {
+            Steps(n) => self.jump_forward(n),
+            TimePeriod(d) => self.jump_duration_forward(d),
+        }
+    }
+}
+
+pub enum StepsOrTimePeriod {
+    Steps(usize),
+    TimePeriod(std::time::Duration),
+}
+
+impl std::str::FromStr for StepsOrTimePeriod {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if let Ok(n) = s.parse::<usize>() {
+            return Ok(StepsOrTimePeriod::Steps(n));
+        }
+        if let Ok(d) = parse_duration::parse(s) {
```

Comment:
I would prefer if we have a manual time parse. I can help write this, what do you think?

Reply:
We were thinking about importing the crate and modifying it: https://github.com/helix-editor/helix/pull/194#discussion_r648326719

Reply:
I'm already planning to add one per archseer's comment. It's better not to duplicate work, there a lot of things that you can contribute to, so my advice would be to wait until mine is in, and improve on it. But if you really feel like writing one yourself right now, go for it :)

Reply:
Yup. Sure, I can do it in a separate pull request as follow up.

Reply:
I added a custom parser, and removed the dependency on parse_duration.